### PR TITLE
Add feature name validation

### PR DESCRIPTION
**Pull request by lhayhurst**
See [this issue](https://github.com/linkedin/feathr/issues/227), "Add validation for feature name".

This PR adds a new method to `feathr_project/feathr/feature.py`, `FeatureBase.validate_feature_name`, that validates the provided feature name against the following rule: 

> Only alphabet, numbers, and '_' are allowed in the name. It can not start with numbers. Note that '.' is NOT ALLOWED

Unit tests are provided for this new function, which is called from the `FeatureBase` constructor before setting the `self.name.`

### CODE REVIEW

File: feathr_project/feathr/feature.py [hunk 1]
```diff
@@ -1,13 +1,20 @@
-from abc import ABC, abstractmethod
```
Comment (lhayhurst) [inline feathr_project/feathr/feature.py:1]: these imports were unused.

File: feathr_project/feathr/feature.py [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
 
 from feathr.dtype import FeatureType
-from feathr.transformation import ExpressionTransformation, Transformation, WindowAggTransformation
-from feathr.typed_key import DUMMY_KEY, TypedKey
 from feathr.frameconfig import HoconConvertible
+from feathr.transformation import ExpressionTransformation, Transformation
+from feathr.typed_key import DUMMY_KEY, TypedKey
+
+
+class FeatureNameValidationError(ValueError):
```
Comment (lhayhurst) [inline feathr_project/feathr/feature.py]: I could not find a standard feathr exception base class when browsing the code, so I created this new exception. Happy to just raise a ValueError or somesuch as well.

File: feathr_project/feathr/feature.py (hunk 1 above)
Comment (hangfei) [reply feathr_project/feathr/feature.py:1]: thanks. please remove them.

File: feathr_project/feathr/feature.py (hunk 2 above)
Comment (hangfei) [reply feathr_project/feathr/feature.py]: There are some new exceptions added for our FeatureRegistry API: https://github.com/linkedin/feathr/blob/main/feathr_project/feathr/api/app/core/feathr_api_exception.py.

For the exception for client, we should create a new one. So we should add the exception you created. Thanks.

File: feathr_project/feathr/feature.py [hunk 3]
```diff
@@ -45,6 +53,27 @@ def __init__(self,
         # self.key could be null, when getting features from registry.
         self.key_alias = [k.key_column_alias for k in self.key if k]
 
+    @classmethod
+    def validate_feature_name(cls, feature_name: str) -> bool:
+        """
+        Only alphabet, numbers, and '_' are allowed in the name.
+                It can not start with numbers. Note that '.' is NOT ALLOWED!
```
Comment (hangfei) [inline feathr_project/feathr/feature.py]: let's add the reasons why "." is not allowed. Some compute engines, like Spark, will consider "." as a operator in the feature name.

File: feathr_project/feathr/feature.py [hunk 4]
```diff
@@ -45,6 +53,27 @@ def __init__(self,
...
+
+        if str.isnumeric(feature_name[0]):
+            raise FeatureNameValidationError(
+                f'Feature name rule violation: feature name cannot start with a number; name={feature_name}')
+
+        feature_validator = re.compile(r'^[a-zA-Z0-9_]+$')
+
+        if not feature_validator.match(feature_name):
+            raise FeatureNameValidationError(
+                f'Feature name rule violation: only letters, numbers, and underscores are allowed in the name; name={feature_name}')
```
Comment (hangfei) [inline feathr_project/feathr/feature.py]: could you try to combine them together and add a documentation for the regex. Let me know if you need help to come up with a regex. I find this useful when i figure out my regex: https://www.regextester.com/

File: feathr_project/test/test_feature_name_validation.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import string
+
+import pytest
+
+from feathr import FeatureBase, FeatureNameValidationError
+
+
+@pytest.mark.parametrize('bad_feature_name',
+                         [None,
+                          ''])
```
Comment (hangfei) [inline feathr_project/test/test_feature_name_validation.py:10]: also add test for "."

File: feathr_project/feathr/feature.py (hunk 2 above)
Comment (lhayhurst) [reply feathr_project/feathr/feature.py]: Done

File: feathr_project/feathr/feature.py (hunk 3 above)
Comment (lhayhurst) [reply feathr_project/feathr/feature.py]: done

File: feathr_project/feathr/feature.py (hunk 4 above)
Comment (lhayhurst) [reply feathr_project/feathr/feature.py]: done

File: feathr_project/test/test_feature_name_validation.py (hunk 5 above)
Comment (lhayhurst) [reply feathr_project/test/test_feature_name_validation.py:10]: This test is actually just testing for empty or None feature names. 

```python
def test_feature_name_fails_on_punctuation_chars():
    for char in set(string.punctuation) - set('_'):
        with pytest.raises(FeatureNameValidationError,  match="only letters, numbers, and underscores are allowed"):
            FeatureBase.validate_feature_name(f"feature_{char}_name")
```

This test checks for "." (which is part of string.punctuation).

**Comment by lhayhurst** [comment]
I'm going to close this PR, as its against my fork, and open it up against the linkedin remote.

**Comment by lhayhurst** [comment]
@hangfei , I've [issued a PR](https://github.com/linkedin/feathr/pull/258) against the linkedin/feathr remote.

**Pull request closed without merging**